### PR TITLE
Add nldas runs

### PR DESCRIPTION
**Pull request by hcorson-dosch-usgs**
Alright - NLDAS models are up and running locally, and are ready to be launched on Tallgrass once this code is approved and merged in. The NLDAS data runs from 1/2/1979 - 4/11/2022. At the moment, per a discussion w/ Jordan, I'm using 1979 for burn-in and 2022 for burn-out, and cropping the output to 1/1/1980-12/31/2021, as that was fastest to set up. In the future we could choose to take the approach we did with the GCM data and add burn-in/out by mirroring the raw data, which would make full use of the available NLDAS data and extend our simulation period.

This should hopefully be a quick review. The substantive code edits were to set up the dates that define the NLDAS model runs `p1_nldas_dates` and build the NLDAS model configuration `p1_nldas_model_config`. The NLDAS models are run using the existing model execution code. The NLDAS results are packaged up using the existing code to process the glm output, with minor edits to make those functions more flexible and update the function documentation.

*Primary edits to set up NLDAS runs:*
* Set up dates that define model run, burn-in, burn-out and how the final data will be cropped
* Build configuration for for NLDAS models
* Build nmls for NLDAS model runs
* Add targets to run NLDAS GLM models (1 model per lake)
* Add targets to extract NLDAS output

*Minor additional edits per our discussion:*
* Rearrange targets into sections
* Add 'driver_type' argument to function to munge model nmls
* Update names of GCM targets to include 'gcm', for clarity
* Switch the GCM configuration column names from 'gcm', 'gcm_start_date', and 'gcm_end_date' to 'driver', 'driver_start_date', and 'driver_end_date' 
* Streamline functions in `'process_glm_output.R'` so that can be called to process GCM or NLDAS GLM output and update function documentation accordingly
* Update p4 targets and scripts to reflect those ^ changes to target and column names

### CODE REVIEW

**Comment by hcorson-dosch-usgs** [comment]
Update - Just tested this code on Tallgrass and then kicked-off the full run. It all ran smoothly and quickly, in about 25 minutes. Since there are now 3569 lakes, there were 3569 runs. 3558/3569 were successful.  Here are the details for the failed runs:
```r
> p2_nldas_glm_uncalibrated_runs %>% filter(glm_success==FALSE) %>% select(site_id, driver, extraction_error, param_sim_stop, max_output_date, glm_code)
# A tibble: 11 × 6
   site_id       driver extraction_error param_sim_stop max_output_date glm_code
   <chr>         <chr>  <chr>            <chr>          <chr>              <int>
 1 nhdhr_114545… NLDAS  NA values        2022-04-11     2022-04-11             0
 2 nhdhr_117648… NLDAS  NA values        2022-04-11     2022-04-11             0
 3 nhdhr_120019… NLDAS  NA values        2022-04-11     2022-04-11             0
 4 nhdhr_120020… NLDAS  NA               2022-04-11     1989-02-23             1
 5 nhdhr_159101… NLDAS  NA values        2022-04-11     2022-04-11             0
 6 nhdhr_597459… NLDAS  NA values        2022-04-11     2022-04-11             0
 7 nhdhr_809976… NLDAS  NA values        2022-04-11     2022-04-11             0
 8 nhdhr_810037… NLDAS  NA values        2022-04-11     2022-04-11             0
 9 nhdhr_810037… NLDAS  NA values        2022-04-11     2022-04-11             0
10 nhdhr_829559… NLDAS  NA values        2022-04-11     2022-04-11             0
11 nhdhr_916842… NLDAS  NA values        2022-04-11     2022-04-11             0
> p2_nldas_glm_uncalibrated_runs %>% filter(glm_success==FALSE) %>% pull(site_id)
 [1] "nhdhr_114545137" "nhdhr_117648183" "nhdhr_120019262" "nhdhr_120020716"
 [5] "nhdhr_159101759" "nhdhr_59745939"  "nhdhr_80997655"  "nhdhr_81003721"
 [9] "nhdhr_81003725"  "nhdhr_82955915"  "nhdhr_91684249"
> dput(p2_nldas_glm_uncalibrated_runs %>% filter(glm_success==FALSE) %>% pull(site_id))
c("nhdhr_114545137", "nhdhr_117648183", "nhdhr_120019262", "nhdhr_120020716",
  "nhdhr_159101759", "nhdhr_59745939", "nhdhr_80997655", "nhdhr_81003721",
  "nhdhr_81003725", "nhdhr_82955915", "nhdhr_91684249")

```

File: 1_prep.R [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
+source('1_prep/src/munge_meteo.R')
 source('1_prep/src/build_model_config.R')
 source('1_prep/src/munge_nmls.R')
 
 p1 <- list(
-  # Define mapping variables
-  tar_target(p1_gcm_names, c('ACCESS', 'GFDL', 'CNRM', 'IPSL', 'MRI', 'MIROC5')),
-  tar_target(p1_gcm_time_periods, c('1981_2000', '2040_2059', '2080_2099')),
+  # Pull in GLM 3 template
+  tar_target(p1_glm_template_nml, '1_prep/in/glm3_template.nml', format = 'file'),
```
Comment (lindsayplatt) [inline 1_prep.R:7]: Since this is an `in/` file, I think you may need to commit?

File: 1_prep/src/munge_nmls.R [hunk 2]
```diff
@@ -21,29 +21,32 @@ adjust_depth_args_nml <- function(nml_args) {
...
+munge_model_nmls <- function(nml_list, base_nml, driver_type) {
   nml_template <- read_nml(base_nml)
   # create the munged nml objects
   nml_objs <- purrr::map(nml_list, function(nml) {
-      # make edits to nml depth-related parameters
-      nml <- purrr::list_modify(nml, !!!adjust_depth_args_nml(nml))
+    # make edits to nml depth-related parameters
+    nml <- purrr::list_modify(nml, !!!adjust_depth_args_nml(nml))
+    # If the nml will note be used for a NLDAS-driven model run
+    if (!(driver_type=='nldas')) {
```
Comment (lindsayplatt) [inline 1_prep/src/munge_nmls.R]:
```suggestion
    if (driver_type!='nldas') {
```

File: 1_prep/src/munge_nmls.R [hunk 3]
```diff
@@ -21,29 +21,32 @@ adjust_depth_args_nml <- function(nml_args) {
   return(nml_edits)
 }
 
-#' @Title Munge default nml lists
-#' @decription Replace the NLDAS filename stored as 'meteo_fl'
-#' with 'NULL', remove used nml parameter 'site_id'
-#' @param nml_list_rds rds file of lake-specific nml parameters
-#' @param site_ids vector of lakes from lake_cell_tile_xwalk
+#' @Title Munge default nml lists for NLDAS runs
```
Comment (lindsayplatt) [inline 1_prep/src/munge_nmls.R]:
```suggestion
#' @Title Munge default nml lists for NLDAS & GCM runs
```

File: 2_run.R [hunk 4]
```diff
@@ -44,13 +46,47 @@ p2 <- list(
...
+  # Group model runs by lake id
+  # Discard the glm diagnostics so they don't trigger rebuilds
+  # even when the export_fl_hash is unchanged
+  # Filter to successful model runs
+  tar_target(
+    p2_nldas_glm_uncalibrated_run_groups,
+    p2_nldas_glm_uncalibrated_runs %>%
+      select(site_id, driver, time_period, driver_start_date, driver_end_date, 
+             export_fl, export_fl_hash, glm_success) %>% 
+      filter(glm_success==TRUE) %>%
```
Comment (lindsayplatt) [inline 2_run.R]:
```suggestion
      filter(glm_success) %>%
```
No need to use `==TRUE` when you are using a value that is already a logical

File: README.md [hunk 5]
```diff
@@ -39,15 +39,15 @@ singularity pull docker://jrossusgs/glm3r:v0.7
 ```
 Here's how to build targets using the Singularity container and the `targets::tar_make_clustermq(target_name, workers=n_workers)` option to build targets in parallel within the Singularity container, with a specified number of workers (up to 72, as Tallgrass has 72 cores per node). The `srun` command will allocate a node and then run the specified command, in this case `Rscript`. Targets will then delegate work out to `n_workers` cores for any parallelizable step that you don't specifically tell it to run in serial.
```
Comment (lindsayplatt) [inline README.md:51]: Do we want to add an example of running the NLDAS GLMs?

**Review by lindsayplatt — APPROVED** [review]
Nice adaptations! Should make it extra easy in the future if we have any new drivers to use. Made a few tiny suggested edits, but also good as-is.

File: 1_prep.R (hunk 1 above)
Comment (hcorson-dosch-usgs) [reply 1_prep.R:7]: Should already have been committed [here](https://github.com/USGS-R/lake-temperature-process-models/commit/17ed224b7f13fd6769e61ac380e077a50c7339b0) -- I just moved the target so it looks like a new addition.

File: 1_prep/src/munge_nmls.R (hunk 3 above)
Comment (hcorson-dosch-usgs) [reply 1_prep/src/munge_nmls.R]: good catch!

**Pull request merged**